README.md
# feedcellar-web

[![Gem Version](https://badge.fury.io/rb/feedcellar-web.svg)](http://badge.fury.io/rb/feedcellar-web)

Feedcellar::Web is a Web interface for [Feedcellar][] that is full-text searchable RSS feed reader and data store.

[Feedcellar]:http://myokoym.net/feedcellar/

## Demo

http://myokoym.net/feedcellar-demo/

## Installation

    $ gem install feedcellar-web

## Usage

### Show help

    $ feedcellar-web

### Show feeds in a web browser

    $ feedcellar-web start [--silent]

Or

    $ rackup

#### Enable cache (using Racknga)

    $ FEEDCELLAR_ENABLE_CACHE=true rackup

## License

Copyright (c) 2013-2015 Masafumi Yokoyama `<myokoym@gmail.com>`

LGPLv2.1 or later.

See 'LICENSE.txt' or 'http://www.gnu.org/licenses/lgpl-2.1' for details.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


NEWS.md
# NEWS

## 0.4.3: 2016-02-17

### Changes

#### Improvements

* Used the latest Feedcellar to support "OR" operator.

#### Fixed

* Don't paginate drilldown source.

## 0.4.2: 2016-02-17

### Changes

#### Improvements

* Always paginate without "page" parameter.
* Paginate in Rroonga.

## 0.4.1: 2016-02-17

### Changes

#### Improvements

* Updated gems.

## 0.4.0: 2016-02-17

### Changes

#### Improvements

* Add space between resources and years.
* Break by years.
* Break by months by years.
* Added /search.json route.
* Supported CORS.
* Don't paginate unless params[:page] is specified.

## 0.3.1: 2015-04-13

Bug fix release for 0.3.0.

### Changes

#### Fixes

* Don't use abstract path for sub URL.
* Hide GitHub ribbon when narrow width.

## 0.3.0: 2015-04-12

### Changes

#### Improvements

* Extracted view files.
* Introduced Bootstrap 3.3.4.
* Added .container.
* Use row grid.
* Extracted #footer.
* footer: Align right.
* Extracted #header.

## 0.2.0: 2015-04-06

### Changes

#### Improvements

* Fixed layouts path.
* Added GitHub ribbon.
* Supported drilldown by year and month.
* Added year drilldown.
* Added month drilldown.
* Showed date of feed.
* Use Feedcellar 0.6.0 or later.

## 0.1.0: 2015-04-05

Initial release.


test/run-test.rb
#!/usr/bin/env ruby
#
# run-test.rb
#
# Copyright (C) 2013-2015  Masafumi Yokoyama <myokoym@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

require "test-unit"
require "test/unit/notify"
require "test/unit/rr"

base_dir = File.expand_path(File.join(File.dirname(__FILE__), ".."))
lib_dir = File.join(base_dir, "lib")
test_dir = File.join(base_dir, "test")
$LOAD_PATH.unshift(lib_dir)
$LOAD_PATH.unshift(test_dir)

exit Test::Unit::AutoRunner.run(true, test_dir)


test/test-app.rb
require "rack/test"

class AppTest < Test::Unit::TestCase
  include Rack::Test::Methods

  def app
    @app ||= Rack::Builder.parse_file("config.ru").first
  end

  def test_index
    get "/"
    assert_true(last_response.ok?)
  end

  def test_search
    get "/search"
    assert_true(last_response.ok?)
  end

  def test_search_json
    get "/search.json"
    assert_true(last_response.ok?)
  end

  def test_registers
    get "/registers.opml"
    assert_true(last_response.ok?)
  end
end


test/test-command.rb
# -*- coding: utf-8 -*-
#
# class CommandTest
#
# Copyright (C) 2015  Masafumi Yokoyama <myokoym@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

require "fileutils"
require "stringio"
require "feedcellar/web/version"
require "feedcellar/web/command"

class CommandTest < Test::Unit::TestCase
  def setup
    @tmpdir = File.join(File.dirname(__FILE__), "tmp")
    FileUtils.rm_rf(@tmpdir)
    FileUtils.mkdir_p(@tmpdir)
    ENV["FEEDCELLAR_HOME"] = @tmpdir
    @command = Feedcellar::Web::Command.new
    @database_dir = @command.database_dir
  end

  def teardown
    FileUtils.rm_rf(@tmpdir)
  end

  def test_version
    s = ""
    io = StringIO.new(s)
    $stdout = io
    @command.version
    assert_equal("#{Feedcellar::Web::VERSION}\n", s)
    $stdout = STDOUT
  end
end


lib/feedcellar/web.rb
# modlue Feedcellar::Web
#
# Copyright (C) 2013-2015  Masafumi Yokoyama <myokoym@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

require "feedcellar/web/version"
require "feedcellar/web/app"


lib/feedcellar/web/command.rb
# class Feedcellar::Web::Command
#
# Copyright (C) 2013-2015  Masafumi Yokoyama <myokoym@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

require "thor"
require "launchy"
require "feedcellar/web"

module Feedcellar
  module Web
    class Command < Thor
      map "-v" => :version

      attr_reader :database_dir

      def initialize(*args)
        super
        default_base_dir = File.join(File.expand_path("~"), ".feedcellar")
        @base_dir = ENV["FEEDCELLAR_HOME"] || default_base_dir
        @database_dir = File.join(@base_dir, "db")
      end

      desc "version", "Show version number."
      def version
        puts Feedcellar::Web::VERSION
      end

      desc "start", "Start web server."
      option :silent, :type => :boolean, :desc => "Don't open in browser"
      def start
        web_server_thread = Thread.new { Feedcellar::Web::App.run! }
        Launchy.open("http://localhost:4567") unless options[:silent]
        web_server_thread.join
      end
    end
  end
end


lib/feedcellar/web/version.rb
# constant Feedcellar::Web::VERSION
#
# Copyright (C) 2015-2016  Masafumi Yokoyama <myokoym@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

module Feedcellar
  module Web
    VERSION = "0.4.4"
  end
end


lib/feedcellar/web/app.rb
# class Feedcellar::Web::App
#
# Copyright (C) 2014-2016  Masafumi Yokoyama <myokoym@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

require "feedcellar"
require "sinatra/base"
require "sinatra/json"
require "sinatra/cross_origin"
require "haml"
require "padrino-helpers"
require "kaminari/sinatra"

module Feedcellar
  module Web
    module PaginationProxy
      def limit_value
        page_size
      end

      def total_pages
        n_pages
      end
    end

    class App < Sinatra::Base
      helpers Kaminari::Helpers::SinatraHelpers
      register Sinatra::CrossOrigin

      get "/" do
        haml :index
      end

      get "/search" do
        search_and_paginate
        haml :index
      end

      get "/search.json" do
        cross_origin
        search_and_paginate
        feeds = @paginated_feeds || @feeds
        json feeds.collect {|feed| feed.attributes }
      end

      get "/registers.opml" do
        content_type :xml
        opml = nil
        GroongaDatabase.new.open(Command.new.database_dir) do |database|
          opml = Opml.build(database.resources.records)
        end
        opml
      end

      helpers do
        def search_and_paginate
          if params[:word]
            words = params[:word].split(" ")
          else
            words = []
          end
          options ||= {}
          options[:resource_id] = params[:resource_id] if params[:resource_id]
          options[:year] = params[:year].to_i if params[:year]
          options[:month] = params[:month].to_i if params[:month]
          @feeds = search(words, options)
          page = (params[:page] || 1).to_i
          size = (params[:n_per_page] || 50).to_i
          @paginated_feeds = @feeds.paginate([["date", :desc]],
                                             page: page,
                                             size: size)
          @paginated_feeds.extend(PaginationProxy)
          @paginated_feeds
        end

        def search(words, options={})
          database = GroongaDatabase.new
          database.open(Command.new.database_dir)
          GroongaSearcher.search(database, words, options)
        end

        def grouping(table)
          key = "resource"
          table.group(key).sort_by {|item| item.n_sub_records }.reverse
        end

        def drilled_url(resource)
          url("/search?resource_id=#{resource._id}&word=#{params[:word]}")
        end

        def drilled_label(resource)
          "#{resource.title} (#{resource.n_sub_records})"
        end

        def groonga_version
          Groonga::VERSION[0..2].join(".")
        end

        def rroonga_version
          Groonga::BINDINGS_VERSION.join(".")
        end
      end
    end
  end
end
